### ISSUE TITLE
relatively slow runtime to find counterfactual explanations

### ISSUE THREAD

**Issue by marcovirgolin** [fat-forensics/fat-forensics#38]
Hi!

I am trying to use FatF's CounterfactualExplainer on the Boston housing data set (among others) for a comparison with other methods.

I set up FatF as you can see here:
https://github.com/marcovirgolin/robust-counterfactuals/blob/dev/robust_cfe/wrappers.py#L201

if you wish to reproduce exactly what I am doing, you can install my repo (branch `dev`) and use this file:
https://github.com/marcovirgolin/robust-counterfactuals/blob/dev/quicktest.sh
(replacing `dice-genetic` with `fatf`)

Some further details: I am passing as ML model a random forest which includes a one-hot encoding pre-processing step whenever .predict is called.

It seems to me it can take a relatively long time (I tried 30 min) to generate counterfactuals. I read that FatF uses grid search.
I guess this means that the search space is sliced into a large number of Cartesian possibilities, and each and every one of them is tried. I guess I could only achieve fast runtime if I use a very coarse grid. 
Am I right about the grid search (and thus why it could be slow), or am I missing something? Is there a max. runtime setting I am missing?

Thank you for your help

**Comment by So-Cool** [comment]
You are correct, it is based on exhaustive search, hence it is computationally expensive. You can try to speed up the search by specifying the list of valid attribute ranges -- only `0` and `1` in your case since you're using one-hot encoding. Otherwise, the explainer treats these features as numerical and may generate redundant search steps. To this end, you can experiment with the `feature_ranges`, `step_sizes` and `default_numerical_step_size` parameters of the [`CounterfactualExplainer`](https://fat-forensics.org/generated/fatf.transparency.predictions.counterfactuals.CounterfactualExplainer.html), but it appears that their default settings already achieve what you're after.

**Comment by marcovirgolin** [comment]
Thank you for the explanation!
